### PR TITLE
💣 Panic recovery middleware

### PR DESCRIPTION
**Pull request by rowanseymour**
and protect against nil pointer panics in type conversion functions

### CODE REVIEW

File: cmd/flowserver/middleware.go [hunk 1]
```diff
@@ -44,23 +40,3 @@ func requestLogger() func(next http.Handler) http.Handler {
...
-func traceErrors() func(next http.Handler) http.Handler {
-	return func(next http.Handler) http.Handler {
-		fn := func(w http.ResponseWriter, r *http.Request) {
-			body := bytes.Buffer{}
-			r.Body = ioutil.NopCloser(io.TeeReader(r.Body, &body))
-			ww := middleware.NewWrapResponseWriter(w, r.ProtoMajor)
-			next.ServeHTTP(ww, r)
-
-			// we are returning a system error of some kind, log the incoming request body
-			if ww.Status() >= 500 && strings.ToLower(r.Method) == "post" {
```
Comment (rowanseymour) [inline cmd/flowserver/middleware.go]: so this never happens - panics end up being converted to 502s at the nginx level

File: excellent/types/text.go [hunk 2]
```diff
@@ -82,5 +82,10 @@ func ToXText(env utils.Environment, x XValue) (XText, XError) {
 		return XTextEmpty, x.(XError)
 	}
 
-	return x.Reduce(env).ToXText(env), nil
+	primitive, isPrimitive := x.(XPrimitive)
+	if isPrimitive {
+		return primitive.ToXText(env), nil
+	}
+
+	return ToXText(env, x.Reduce(env))
```
Comment (rowanseymour) [inline excellent/types/text.go]: `Reduce` could return a nil in the case of unset field value etc

File: cmd/flowserver/middleware.go (hunk 1 above)
Comment (nicpottier) [reply cmd/flowserver/middleware.go]: are we still logging the request somewhere? Seems `Recoverer` actually traps the panic then changes the response to a 500 so maybe we were just missing that piece?

Comment (rowanseymour) [reply cmd/flowserver/middleware.go]: Maybe - but I've now changed it to use a combination of raven's own recovery middleware (not made for chi) and chi's own implementation. The recovery middleware comes before the request logger middleware - so the former recovers from the panic, logs it to sentry along with the request, converts it to 500 response - the latter will then log that. In theory. Not the easiest thing to test.

**Pull request merged**